src/MOI_wrapper.jl

```diff
@@ -50,22 +59,40 @@ mutable struct Optimizer <: MOI.AbstractOptimizer
     end
 end
 
+"Release references and free memory."
+function free_scip(o::Optimizer)
+    # Avoid double-free (SCIP will set the pointers to NULL).
+    if o.inner.scip[] != C_NULL
+        for c in values(o.inner.conss)
+            @SCIP_CALL SCIPreleaseCons(o.inner, c)
+        end
+        for v in values(o.inner.vars)
+            @SCIP_CALL SCIPreleaseVar(o.inner, v)
+        end
+        # only o.inner.scip is GC-protected during ccall!
+        GC.@preserve o begin
+            @SCIP_CALL SCIPfree(o.inner.scip)
+        end
+    end
+    @assert o.inner.scip[] == C_NULL
+end
+
 # Protect Optimizer from GC for ccall with Ptr{SCIP_} argument.
-Base.unsafe_convert(::Type{Ptr{SCIP_}}, o::Optimizer) = o.mscip.scip[]
+Base.unsafe_convert(::Type{Ptr{SCIP_}}, o::Optimizer) = o.inner.scip[]
```

Comment:
Please read https://jump.dev/MathOptInterface.jl/stable/manual/implementing/#Optimizer-objects-for-C-solvers

Reply:
Note that the `[]` operator is used here, because the [SCIPfree](https://www.scipopt.org/doc-7.0.2/html/group__GeneralSCIPMethods.php#gad8ae80418f738b5b10e3b0fd08782bed) function expects and argument `SCIP ** scip` (which will be set to `NULL`). That is, we don't store `SCIP*` but actually `SCIP**`.

Reply:
I've removed the `@preserve` in line 73 and added `Base.cconvert(::Type{Ptr{SCIP_}}, o::Optimizer) = o` right before the `unsafe_convert` in line 81 but I don't think that's quite correct. Where have I gone wrong?

Reply:
> but I don't think that's quite correct

what do you mean specifically?

Reply:
I'm getting 

> error in running finalizer: AssertionError(msg="o.inner.scip[] == C_NULL")
create and semi-manual free: Test Failed at /home/erik/scipjl/SCIP.jl/test/scip_data.jl:26
  Expression: o.inner.scip[] == C_NULL
   Evaluated: Ptr{Nothing} @0x00000000024680c0 == Ptr{Nothing} @0x0000000000000000
Stacktrace:
 [1] top-level scope at /home/erik/scipjl/SCIP.jl/test/scip_data.jl:26
 [2] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.5/Test/src/Test.jl:1115
 [3] top-level scope at /home/erik/scipjl/SCIP.jl/test/scip_data.jl:23

Reply:
Found it, I misplaced a line.